blogs/treewidth.md

```diff
@@ -0,0 +1,385 @@
+3. Bags containing the same vertex have to be connected in the tree.
+
+An example of tree decomposition is shown in the following figure:
+
+![](/assets/treewidth_figs/treedecomposition.png)
+
+where the left graph is the original graph and the right one is the tree decomposition of the graph.
+
+All the nodes of the tree are called **tree bags**, and the intersection of two bags is called a **separator**.
+The width of a tree decomposition is the size of the largest bag minus one, thus the width of the example above is $2$.
+Clearly, one graph can have multiple tree decompositions, and the **tree width** of a graph is the minimum width of all possible tree decompositions.
+For graphs with weights on vertices, we define the width as sum of weights of the vertices in the bags minus one.
+
+### From Tree Decomposition to Contraction Order
+
+Then how tree decomposition and tree width are related to the contraction order of the tensor network?
+In this sub-section, we will answer the following questions: how to get a contraction order from a tree decomposition? And why the contraction order is optimal if the tree width is minimal?
+
+For the first question, according to the definition of tree decomposition, we can obtain vertices elimination order by traversing the tree decomposition in a bottom-up manner. 
+For each node of the tree decomposition, a vertex can be eliminated if it is in the bag of the node and not in the bags of its parent node.
+The elimination order can then be used to determine the contraction order of the tensor network, where two tensors are contracted if they share a common index and the index is eliminated.
+An example of contraction order is shown in the following figure:
+
+![](/assets/treewidth_figs/elimination.png)
+
+where the tree decomposition gives the elimination order $a, b, d, f, c, h, e, g, i$, and this lead to a contraction order $(((((A, B), C), F), D), E)$.
+Of course, the contraction order is not unique by selecting different node as the root of the decomposition tree, but the space complexity of the contraction is the same for all possible contraction orders.
+
+For the second question, we have the following two reasons:
+
+1. Width of the tree decomposition is an upper bound of the space complexity of the corresponding contraction.
+2. Total dimension of the indices in the same tree bag is the time complexity of the corresponding step of the contraction.
+
+For example, in the first step of the construction above, the index $a$ is eliminated, and the corresponding operations is
+$$A'_{bcde} = \sum_{a} A_{abc} B_{cde}$$
+where the space complexity is bounded by width and the time complexity is given by the dimension of the indices in bag.
+
+Thus, finding the tree decomposition with minimal width is equivalent to finding the optimal contraction order of the tensor network, with a bounded space complexity and minimal time complexity.
+
+## The Bouchitté–Todinca Algorithm for Tree Width
```

Comment:
I did not yet go through this section in detail, as there are still some concepts and definitions in the beginning that I do not fully understand; see follow up comments: